File: lib/xai/grok-client.ts
/**
 * xAI Grok API Client with Function Calling Support (TypeScript)
 * Integrates Grok-4 with custom tools for THEOS Sovereign OS
 */

import { xai } from '@ai-sdk/xai';
import { streamText, tool, stepCountIs, generateText } from 'ai';
import { z } from 'zod';

// ============================================================================
// Tool Definitions
// ============================================================================

const getCurrentTemperature = tool({
  description: 'Get the current temperature in a given location',
  inputSchema: z.object({
    location: z
      .string()
      .describe('The city and state, e.g. San Francisco, CA'),
    unit: z
      .enum(['celsius', 'fahrenheit'])
      .default('fahrenheit')
      .describe('Temperature unit'),
  }),
  execute: async ({ location, unit }) => {
    const temperature = unit === 'fahrenheit' ? 59 : 15;
    return {
      location,
      temperature,
      unit,
    };
  },
});

const getCurrentCeiling = tool({
  description: 'Get the current cloud ceiling in a given location',
  inputSchema: z.object({
    location: z
      .string()
      .describe('The city and state, e.g. San Francisco, CA'),
  }),
  execute: async ({ location }) => {
    return {
      location,
      ceiling: 15000,
      ceiling_type: 'broken',
      unit: 'ft',
    };
  },
});

// ============================================================================
// THEOS-Specific Tools
// ============================================================================

const queryOracle = tool({
  description: 'Query the THEOS Final Oracle for a covenant vector address (DAUS, ALIMA, TREASURY_OF_LIGHT, COVENANT_ROOT, ARCHITECT_MIND, LEDGER_OWNER, or CHARIOT_STELLAR)',
  inputSchema: z.object({
    vector: z
      .enum(['DAUS', 'ALIMA', 'TREASURY_OF_LIGHT', 'COVENANT_ROOT', 'ARCHITECT_MIND', 'LEDGER_OWNER', 'CHARIOT_STELLAR'])
      .describe('Covenant vector name'),
    chainId: z
      .number()
      .default(42161)
      .describe('Chain ID (default: 42161 for Arbitrum One)'),
  }),
  execute: async ({ vector, chainId }) => {
    // Canonical Oracle addresses
    const ORACLE_ADDRESSES: Record<string, string> = {
      '42161': '0x8BCbC66A5bb808A8871F667f2dd92a017B3DaFAC', // Arbitrum One
      '1': '0x8BCbC66A5bb808A8871F667f2dd92a017B3DaFAC',     // Ethereum Mainnet (if deployed)
    };

    // Canonical vectors
    const VECTORS: Record<string, string> = {
      'DAUS': '0x8BCbC66a5bb808A8871F667f2Dd92a017B3DaFAC',
      'ALIMA': '0xC775BF1118f44B8a72268aFacF8F7F2ef53A6D24',
      'TREASURY_OF_LIGHT': '0xb4C173AaFe428845f0b96610cf53576121BAB221',
      'COVENANT_ROOT': '0xD98CF268718e925D53314662e0478EE13517FD54',
      'ARCHITECT_MIND': '0x3BBa654A3816A228284E3e0401Cff4EA6dFc5cea',
      'LEDGER_OWNER': '0x3df07977140Ad97465075129C37Aec7237d74415',
      'CHARIOT_STELLAR': 'GCGCUGIOCJLRRP3JSRZEBOAMIFW5EM3WZUT5S3DXVC7I44N5I7FN5C2F',
    };

    return {
      vector,
      address: VECTORS[vector],
      chainId,
      oracleAddress: ORACLE_ADDRESSES[chainId.toString()] || 'Not deployed on this chain',
      status: 'canonical',
    };
  },
});

const getSafeStatus = tool({
  description: 'Get the status of a Safe wallet, including Treasury of Light information',
  inputSchema: z.object({
    safeAddress: z
      .string()
      .describe('Safe wallet address'),
    chainId: z
      .number()
      .default(42161)
      .describe('Chain ID (default: 42161 for Arbitrum One)'),
  }),
  execute: async ({ safeAddress, chainId }) => {
    const TREASURY_OF_LIGHT = '0xb4C173AaFe428845f0b96610cf53576121BAB221';
    const isTreasury = safeAddress.toLowerCase() === TREASURY_OF_LIGHT.toLowerCase();

    return {
      safeAddress,
      chainId,
      isTreasuryOfLight: isTreasury,
      threshold: isTreasury ? 2 : 'unknown',
      owners: isTreasury
        ? [
            '0x3BBa654A3816A228284E3e0401Cff4EA6dFc5cea', // tig08bitties.uni.eth
            '0x3df07977140Ad97465075129C37Aec7237d74415', // Ledger Flex
          ]
        : 'unknown',
      status: isTreasury ? 'active' : 'unknown',
    };
  },
});

const verifyAddress = tool({
  description: 'Verify an Ethereum address against canonical covenant vectors',
  inputSchema: z.object({
    address: z
      .string()
      .describe('Ethereum address to verify'),
    chainId: z
      .number()
      .describe('Chain ID to check on'),
  }),
  execute: async ({ address, chainId }) => {
    const addressLower = address.toLowerCase();

    // Canonical addresses
    const CANONICAL: Record<string, string> = {
      '0x8bcc66a5bb808a8871f667f2dd92a017b3dafac': 'DAUS',
      '0xc775bf1118f44b8a72268afacf8f7f2ef53a6d24': 'ALIMA',
      '0xb4c173aafe428845f0b96610cf53576121bab221': 'TREASURY_OF_LIGHT',
      '0xd98cf268718e925d53314662e0478ee13517fd54': 'COVENANT_ROOT',
      '0x3bba654a3816a228284e3e0401cff4ea6dfc5cea': 'ARCHITECT_MIND',
      '0x3df07977140ad97465075129c37aec7237d74415': 'LEDGER_OWNER',
      '0x3b34c30f51fe6e276530aacb8f4d877e9893356f': 'SAFE_DEPLOYER_AGENT',
    };

    const vectorName = CANONICAL[addressLower];

    return {
      address,
      chainId,
      isCanonical: vectorName !== undefined,
      vectorName: vectorName || null,
      verificationStatus: vectorName ? 'verified' : 'unknown',
    };
  },
});

// ============================================================================
// Grok Client Class
// ============================================================================

export class GrokClient {
  private apiKey: string;

  constructor(apiKey?: string) {
    this.apiKey = apiKey || process.env.XAI_API_KEY || '';
    if (!this.apiKey) {
      throw new Error('XAI_API_KEY not provided and not found in environment');
    }
  }

  /**
   * Stream text with tools
   */
  async streamText(
    prompt: string,
    systemMessage?: string,
    options?: {
      maxSteps?: number;
      temperature?: number;
      includeTheosTools?: boolean;
    }
  ) {
    const tools: Record<string, any> = {
      getCurrentTemperature,
      getCurrentCeiling,
    };

    if (options?.includeTheosTools) {
      tools.queryOracle = queryOracle;
      tools.getSafeStatus = getSafeStatus;
      tools.verifyAddress = verifyAddress;
    }

    const result = streamText({
      model: xai('grok-4-latest', {
        apiKey: this.apiKey,
      }),
      tools,
      stopWhen: stepCountIs(options?.maxSteps || 5),
      system: systemMessage,
      prompt,
      temperature: options?.temperature || 0,
    });

    return result;
  }

  /**
   * Generate text with tools (non-streaming)
   */
  async generateText(
    prompt: string,
    systemMessage?: string,
    options?: {
      maxSteps?: number;
      temperature?: number;
      includeTheosTools?: boolean;
    }
  ) {
    const tools: Record<string, any> = {
      getCurrentTemperature,
      getCurrentCeiling,
    };

    if (options?.includeTheosTools) {
      tools.queryOracle = queryOracle;
      tools.getSafeStatus = getSafeStatus;
      tools.verifyAddress = verifyAddress;
    }

    const result = await generateText({
      model: xai('grok-4-latest', {
        apiKey: this.apiKey,
      }),
      tools,
      maxSteps: options?.maxSteps || 5,
      system: systemMessage,
      prompt,
      temperature: options?.temperature || 0,
    });

    return result;
  }

  /**
   * Process stream and handle tool calls
   */
  async processStream(
    stream: any,
    onText?: (text: string) => void,
    onToolCall?: (toolName: string, input: any) => void,
    onToolResult?: (toolName: string, output: any) => void
  ) {
    let fullText = '';

    for await (const chunk of stream.fullStream) {
      switch (chunk.type) {
        case 'text-delta':
          if (onText) {
            onText(chunk.text);
          }
          fullText += chunk.text;
          break;
        case 'tool-call':
          if (onToolCall) {
            onToolCall(chunk.toolName, chunk.args);
          }
          break;
        case 'tool-result':
          if (onToolResult) {
            onToolResult(chunk.toolName, chunk.result);
          }
          break;
      }
    }

    return fullText;
  }
}

// ============================================================================
// Example Usage
// ============================================================================

if (require.main === module) {
  (async () => {
    const client = new GrokClient();

    // Example 1: Stream with THEOS tools
    console.log('Example 1: Querying THEOS Oracle...\n');
    const stream1 = await client.streamText(
      "What's the address of the TREASURY_OF_LIGHT?",
      'You are a helpful assistant for the THEOS Sovereign OS.',
      { includeTheosTools: true }
    );

    await client.processStream(
      stream1,
      (text) => process.stdout.write(text),
      (toolName, input) => console.log(`\n[Tool Call: ${toolName}]`, input),
      (toolName, output) => console.log(`\n[Tool Result: ${toolName}]`, output)
    );

    console.log('\n\n');

    // Example 2: Simple temperature query
    console.log('Example 2: Temperature query...\n');
    const stream2 = await client.streamText(
      "What's the temperature like in Sioux Falls, SD?",
      'You are a helpful weather assistant.'
    );

    await client.processStream(stream2, (text) => process.stdout.write(text));

    console.log('\n\n');

    // Example 3: Generate text (non-streaming)
    console.log('Example 3: Non-streaming generation...\n');
    const result = await client.generateText(
      'What is the DAUS vector address?',
      'You are a helpful assistant for the THEOS Sovereign OS.',
      { includeTheosTools: true }
    );

    console.log(result.text);
  })();
}

File: lib/xai/grok-client.py
"""
xAI Grok API Client with Function Calling Support
Integrates Grok-4 with custom tools for THEOS Sovereign OS
"""

import json
from typing import Literal, Dict, Any, List, Optional, Callable
from pydantic import BaseModel, Field
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

try:
    from xai import XAI
except ImportError:
    XAI = None
    print("Warning: xAI SDK not installed. Install with: pip install xai")


# ============================================================================
# Pydantic Models for Tool Definitions
# ============================================================================

class TemperatureRequest(BaseModel):
    """Request model for temperature queries"""
    location: str = Field(description="The city and state, e.g. San Francisco, CA")
    unit: Literal["celsius", "fahrenheit"] = Field(
        "fahrenheit", description="Temperature unit"
    )


class CeilingRequest(BaseModel):
    """Request model for cloud ceiling queries"""
    location: str = Field(description="The city and state, e.g. San Francisco, CA")


# ============================================================================
# Tool Functions
# ============================================================================

def get_current_temperature(request: TemperatureRequest) -> Dict[str, Any]:
    """
    Get the current temperature in a given location
    
    Args:
        request: TemperatureRequest with location and unit
        
    Returns:
        Dictionary with location, temperature, and unit
    """
    temperature = 59 if request.unit.lower() == "fahrenheit" else 15
    return {
        "location": request.location,
        "temperature": temperature,
        "unit": request.unit,
    }


def get_current_ceiling(request: CeilingRequest) -> Dict[str, Any]:
    """
    Get the current cloud ceiling in a given location
    
    Args:
        request: CeilingRequest with location
        
    Returns:
        Dictionary with location, ceiling, ceiling_type, and unit
    """
    return {
        "location": request.location,
        "ceiling": 15000,
        "ceiling_type": "broken",
        "unit": "ft",
    }


# ============================================================================
# Tool Definitions Generator
# ============================================================================

def generate_tool_definitions() -> List[Dict[str, Any]]:
    """
    Generate tool definitions from Pydantic models
    
    Returns:
        List of tool definition dictionaries
    """
    get_current_temperature_schema = TemperatureRequest.model_json_schema()
    get_current_ceiling_schema = CeilingRequest.model_json_schema()
    
    return [
        {
            "type": "function",
            "function": {
                "name": "get_current_temperature",
                "description": "Get the current temperature in a given location",
                "parameters": get_current_temperature_schema,
            }
        },
        {
            "type": "function",
            "function": {
                "name": "get_current_ceiling",
                "description": "Get the current cloud ceiling in a given location",
                "parameters": get_current_ceiling_schema,
            }
        },
    ]


# Tools mapping for function execution
TOOLS_MAP: Dict[str, Callable] = {
    "get_current_temperature": get_current_temperature,
    "get_current_ceiling": get_current_ceiling,
}


# ============================================================================
# Grok Client Class
# ============================================================================

class GrokClient:
    """
    xAI Grok API Client with function calling support
    """
    
    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize Grok client
        
        Args:
            api_key: xAI API key (defaults to XAI_API_KEY env var)
        """
        self.api_key = api_key or os.getenv("XAI_API_KEY")
        if not self.api_key:
            raise ValueError("XAI_API_KEY not provided and not found in environment")
        
        if XAI is None:
            raise ImportError("xAI SDK not installed. Install with: pip install xai")
        
        self.client = XAI(api_key=self.api_key)
        self.tool_definitions = generate_tool_definitions()
        self.tools_map = TOOLS_MAP.copy()
    
    def add_tool(self, name: str, function: Callable, model: BaseModel, description: str):
        """
        Add a custom tool to the client
        
        Args:
            name: Tool function name
            function: Callable function
            model: Pydantic model for parameters
            description: Tool description
        """
        schema = model.model_json_schema()
        tool_def = {
            "type": "function",
            "function": {
                "name": name,
                "description": description,
                "parameters": schema,
            }
        }
        self.tool_definitions.append(tool_def)
        self.tools_map[name] = function
    
    def chat(
        self,
        messages: List[Dict[str, str]],
        model: str = "grok-4-latest",
        tool_choice: str = "auto",
        temperature: float = 0,
        stream: bool = False
    ) -> Any:
        """
        Send a chat completion request
        
        Args:
            messages: List of message dictionaries
            model: Model name (default: "grok-4-latest")
            tool_choice: Tool choice mode (default: "auto")
            temperature: Sampling temperature (default: 0)
            stream: Whether to stream responses (default: False)
            
        Returns:
            Chat completion response
        """
        return self.client.chat.completions.create(
            model=model,
            messages=messages,
            tools=self.tool_definitions,
            tool_choice=tool_choice,
            temperature=temperature,
            stream=stream
        )
    
    def execute_tool_call(self, tool_call: Any) -> Dict[str, Any]:
        """
        Execute a tool call from Grok's response
        
        Args:
            tool_call: Tool call object from Grok response
            
        Returns:
            Tool response dictionary
        """
        function_name = tool_call.function.name
        
        if function_name not in self.tools_map:
            return {
                "error": f"Function {function_name} not found",
                "tool_call_id": tool_call.id
            }
        
        try:
            function_args = json.loads(tool_call.function.arguments)
            
            # Get the appropriate Pydantic model for validation
            if function_name == "get_current_temperature":
                request = TemperatureRequest(**function_args)
                result = self.tools_map[function_name](request)
            elif function_name == "get_current_ceiling":
                request = CeilingRequest(**function_args)
                result = self.tools_map[function_name](request)
            else:
                # Generic call for custom tools
                result = self.tools_map[function_name](**function_args)
            
            return {
                "content": json.dumps(result),
                "tool_call_id": tool_call.id
            }
        except Exception as e:
            return {
                "error": str(e),
                "tool_call_id": tool_call.id
            }
    
    def chat_with_tools(
        self,
        user_message: str,
        system_message: Optional[str] = None,
        model: str = "grok-4-latest",
        max_iterations: int = 5
    ) -> str:
        """
        Complete chat flow with automatic tool calling
        
        Args:
            user_message: User's message
            system_message: Optional system message
            model: Model name
            max_iterations: Maximum tool call iterations
            
        Returns:
            Final response content
        """
        messages = []
        
        if system_message:
            messages.append({"role": "system", "content": system_message})
        
        messages.append({"role": "user", "content": user_message})
        
        iteration = 0
        while iteration < max_iterations:
            # Get response from Grok
            response = self.chat(messages, model=model)
            assistant_message = response.choices[0].message
            
            # Append assistant message
            messages.append({
                "role": assistant_message.role,
                "content": assistant_message.content,
                "tool_calls": [
                    {
                        "id": tc.id,
                        "type": tc.type,
                        "function": {
                            "name": tc.function.name,
                            "arguments": tc.function.arguments
                        }
                    } for tc in (assistant_message.tool_calls or [])
                ] if assistant_message.tool_calls else None
            })
            
            # Check if there are tool calls
            if assistant_message.tool_calls:
                for tool_call in assistant_message.tool_calls:
                    tool_result = self.execute_tool_call(tool_call)
                    messages.append({
                        "role": "tool",
                        "content": tool_result.get("content", json.dumps(tool_result)),
                        "tool_call_id": tool_result.get("tool_call_id")
                    })
                iteration += 1
            else:
                # No more tool calls, return final response
                return assistant_message.content or ""
        
        # Max iterations reached
        return messages[-1].get("content", "Maximum iterations reached")


# ============================================================================
# Example Usage
# ============================================================================

if __name__ == "__main__":
    # Initialize client
    client = GrokClient()
    
    # Example 1: Simple chat with tools
    response = client.chat_with_tools(
        user_message="What's the temperature like in San Francisco?",
        system_message="You are a helpful weather assistant."
    )
    print("Response:", response)
    
    # Example 2: Manual tool calling
    messages = [{"role": "user", "content": "What's the temperature in Sioux Falls, SD?"}]
    response = client.chat(messages)
    
    if response.choices[0].message.tool_calls:
        for tool_call in response.choices[0].message.tool_calls:
            result = client.execute_tool_call(tool_call)
            print("Tool result:", result)
